[WebGPU] Selectively emitting bufferLength allows for mismatched bind groups

tadeuzagallo opened the pull request:
#### 661c534f55532c756caea32a63c21beec459df11
<pre>
[WebGPU] Selectively emitting bufferLength allows for mismatched bind groups
<a href="https://bugs.webkit.org/show_bug.cgi?id=291533">https://bugs.webkit.org/show_bug.cgi?id=291533</a>
<a href="https://rdar.apple.com/149183515">rdar://149183515</a>

Reviewed by Mike Wyrzykowski.

Automatically generated bind groups can be used in place of other compatible
bind groups of the same pipeline, but to ensure compatibility we must emit all
buffer lengths. To do that we force the generation of all length variables as
long as the buffer isn&apos;t an external texture (to match the API behavior) and
the variable associated with the length has been used.

* LayoutTests/fast/webgpu/fuzz-149183515-reduced-expected.txt: Added.
* LayoutTests/fast/webgpu/fuzz-149183515-reduced.html: Added.
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::collectGlobals):
(WGSL::bindingTypeForPrimitive):
(WGSL::bindingTypeForType):
(WGSL::isExternalTexture):
(WGSL::RewriteGlobalVariables::insertStructs):

Canonical link: <a href="https://commits.webkit.org/293752@main">https://commits.webkit.org/293752@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c67e33fd95b608d04ebb59763870fee0ebbf3ec6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99834 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19479 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9759 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104960 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50412 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19784 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75991 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33081 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102841 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15078 "Found 1 new test failure: fast/forms/ios/file-upload-panel-dismiss-when-view-removed-from-window.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90145 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56356 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14883 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8132 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49784 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84802 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8217 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107320 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26945 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19675 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84945 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27307 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86350 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84471 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21445 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29146 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6861 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20752 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26881 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32092 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26692 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30009 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28253 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on LayoutTests/fast/webgpu/fuzz-149183515-reduced.html:
```diff
@@ -0,0 +1,163 @@
+    computePassEncoder8.end();
+    let commandBuffer1 = commandEncoder2.finish();
+    device0.queue.submit([commandBuffer1]);
+    await device0.queue.onSubmittedWorkDone();
```
`videoFrame1.close()` should fix the test output

tadeuzagallo committed:
[WebGPU] Selectively emitting bufferLength allows for mismatched bind groups
https://bugs.webkit.org/show_bug.cgi?id=291533
rdar://149183515

Reviewed by Mike Wyrzykowski.

Automatically generated bind groups can be used in place of other compatible
bind groups of the same pipeline, but to ensure compatibility we must emit all
buffer lengths. To do that we force the generation of all length variables as
long as the buffer isn't an external texture (to match the API behavior) and
the variable associated with the length has been used.

* LayoutTests/fast/webgpu/fuzz-149183515-reduced-expected.txt: Added.
* LayoutTests/fast/webgpu/fuzz-149183515-reduced.html: Added.
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::collectGlobals):
(WGSL::bindingTypeForPrimitive):
(WGSL::bindingTypeForType):
(WGSL::isExternalTexture):
(WGSL::RewriteGlobalVariables::insertStructs):

Canonical link: https://commits.webkit.org/293752@main
